examples/examples/tower_http.rs

```diff
@@ -0,0 +1,130 @@
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use hyper::body::Bytes;
+use hyper::service::make_service_fn;
+use hyper::Server;
+use std::convert::Infallible;
+use std::iter::once;
+use std::net::SocketAddr;
+use std::time::{Duration, Instant};
+use tower_http::sensitive_headers::SetSensitiveRequestHeadersLayer;
+use tower_http::trace::{DefaultMakeSpan, DefaultOnResponse, TraceLayer};
+use tower_http::LatencyUnit;
+
+use jsonrpsee::core::client::ClientT;
+use jsonrpsee::core::middleware::{self, Headers, Params};
+use jsonrpsee::http_client::HttpClientBuilder;
+use jsonrpsee::http_server::{HttpServerBuilder, RpcModule};
+
+#[derive(Clone)]
+struct Timings;
+
+impl middleware::HttpMiddleware for Timings {
+	type Instant = Instant;
+
+	fn on_request(&self, remote_addr: SocketAddr, headers: &Headers) -> Self::Instant {
+		println!("[Middleware::on_request] remote_addr {}, headers: {:?}", remote_addr, headers);
+		Instant::now()
+	}
+
+	fn on_call(&self, name: &str, params: Params, kind: middleware::MethodKind) {
+		println!("[Middleware::on_call] method: '{}', params: {:?}, kind: {}", name, params, kind);
+	}
+
+	fn on_result(&self, name: &str, success: bool, started_at: Self::Instant) {
+		println!("[Middleware::on_result] '{}', worked? {}, time elapsed {:?}", name, success, started_at.elapsed());
+	}
+
+	fn on_response(&self, result: &str, started_at: Self::Instant) {
+		println!("[Middleware::on_response] result: {}, time elapsed {:?}", result, started_at.elapsed());
+	}
+}
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+	tracing_subscriber::FmtSubscriber::builder()
+		.with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
+		.try_init()
+		.expect("setting default subscriber failed");
+
+	let addr = run_server().await?;
+	let url = format!("http://{}", addr);
+	println!("[main]: URL {:?}", url);
+
+	let client = HttpClientBuilder::default().build(&url)?;
+	let response: String = client.request("say_hello", None).await?;
+	println!("[main]: response: {:?}", response);
+	let _response: Result<String, _> = client.request("unknown_method", None).await;
+	let _ = client.request::<String>("say_hello", None).await?;
+
+	// Make the same request again.
+	let _ = client.request::<String>("say_hello", None).await?;
+
+	Ok(())
+}
+
+async fn run_server() -> anyhow::Result<SocketAddr> {
+	let addr = SocketAddr::from(([127, 0, 0, 1], 9935));
+
+	let make_service = make_service_fn(move |_| {
+		async move {
+			let mut module = RpcModule::new(());
+			module.register_method("say_hello", |_, _| Ok("lo")).unwrap();
+
+			println!("[run_server]: Creating RPC service");
+			let rpc_svc = HttpServerBuilder::new().set_middleware(Timings).to_service(module).unwrap();
+
+			println!("[run_server]: Tower builder");
+			let tower_svc = tower::ServiceBuilder::new()
```

Comment:
Is there a reason why we shouldn't enable this be the default on the HTTP server i.e. `tower middleware`?
I reckon that most folks would find it useful to debug the entire HTTP request/response.

However, the logging/middleware stuff becomes quite complicated because we got:

1. RpcMiddleware
2. Tracing of RPC calls
3. HTTP middleware

It's probably fine but a bit overwhelming to understand the difference :P

Reply:
I agree with the middleware being confusing bit. Some of it just a naming thing; calling the logging stuff Middleware was probably a mistake, since it isn't really; it's just some basic hooks to log details from and can't effect any change.

There are different sorts of middleware support that are interesting:

- HTTP Service (like this PR; just an obvious thing to support since tower is part of hyper already. This is the lowest hanging fruit I guess. CORS could lilve here, and we keep more basic IP filtering in jsonrpsee itself)
- WS Service (use tower service trait and allow things to be put in front of each request/response pair?)
- Request/Response Service (Like, regardless of HTTP or WS, a general service which would receive requests/responses and be able to act on/modify them? Maybe this is strictly just better than a WS specific thing?)
- The "manual" approach; allow some custom param `T` to be passed into call/subscription handlers so that the user has a way to get custom information in from the outside (worst case by manually hooking up an http service and manually calling things on RpcModule with said custom state; maximum flexibility but hardest to make use of).